databricks/koalas/frame.py

```diff
@@ -5306,6 +5307,148 @@ def melt(self, id_vars=None, value_vars=None, var_name='variable',
+        0  1.0  3.0
+        1  2.0  4.0
+        2  4.0  2.0
+        3  3.0  1.0
+        """
+        if method not in ['average', 'min', 'max', 'first', 'dense']:
+            msg = "method must be one of 'average', 'min', 'max', 'first', 'dense'"
+            raise ValueError(msg)
+
+        if na_option not in ['keep', 'top', 'bottom']:
+            msg = "na_option must be one of 'keep', 'top', 'bottom'"
+            raise ValueError(msg)
+
+        if ascending:
+            asc_func = spark.functions.asc
+        else:
+            asc_func = spark.functions.desc
+
+        index_column = self._internal.index_columns[0]
+        data_columns = self._internal.data_columns
+        sdf = self._sdf
+
+        for column_name in data_columns:
+            if na_option == 'top':
+                if not ascending:
+                    fill_value = sys.maxsize
+            elif na_option == 'bottom':
+                if ascending:
+                    fill_value = sys.maxsize
+            elif na_option == 'keep':
+                if ascending:
+                    fill_value = sys.maxsize
+                else:
+                    fill_value = -sys.maxsize
+
+            if 'fill_value' in locals():
+                sdf = sdf.withColumn(column_name + '_tmp', F.col(column_name))
+                sdf = sdf.fillna({column_name: fill_value})
+
+            if method == 'first':
+                window = Window.orderBy(asc_func(column_name), asc_func(index_column))\
+                    .rowsBetween(Window.unboundedPreceding, Window.currentRow)
+                sdf = sdf.withColumn(column_name, F.row_number().over(window))
+            elif method == 'dense':
+                window = Window.orderBy(asc_func(column_name))\
+                    .rowsBetween(Window.unboundedPreceding, Window.currentRow)
+                sdf = sdf.withColumn(column_name, F.dense_rank().over(window))
+            else:
+                if method == 'average':
+                    stat_func = F.mean
+                elif method == 'min':
+                    stat_func = F.min
+                elif method == 'max':
+                    stat_func = F.max
+                window = Window.orderBy(asc_func(column_name))\
+                    .rowsBetween(Window.unboundedPreceding, Window.currentRow)
+                sdf = sdf.withColumn('rank', F.row_number().over(window))
+                window = Window.partitionBy(column_name)\
+                    .rowsBetween(Window.unboundedPreceding, Window.unboundedFollowing)
+                sdf = sdf.withColumn(column_name, stat_func(F.col('rank')).over(window))
+
+            if na_option == 'keep':
+                sdf = sdf.withColumn(column_name,
+                                     F.when(sdf[column_name + '_tmp'].isNull(), F.lit(None))
+                                     .otherwise(sdf[column_name]))
+
+        return DataFrame(self._internal.copy(sdf=sdf.select(self._internal.columns)))\
+            .astype(np.float64)
```

Comment:
Why do we need to cast it, out of curiosity?

Reply:
some spark rank function will return int but pandas.rank() return float all the time.